query/outputnode.go

```diff
@@ -812,11 +812,50 @@ func (enc *encoder) merge(parent, child []fastJsonNode) ([]fastJsonNode, error)
 			if paCopy == nil {
 				paCopy = caCopy
 			} else {
-				temp := paCopy
+				// This code merge child and parent lists such that nodes with
+				// same attribute id's comes together in merged list
+				tempPa := paCopy
+				// unmatchedNodesStartPtr and unmatchedNodesEndPtr stores start
+				// and end pointer of whose attribute id's  doesn't match with
+				// attribute id's of some nodes in parent list
+				var unmatchedNodesStartPtr fastJsonNode
+				var unmatchedNodesEndPtr fastJsonNode
+				for caCopy != nil {
+					var exist bool
+					caCopyNext := caCopy.next
+					for paCopy != nil {
+						// Merge child nodes in parent list whose attribute id's matched
+						// with attribute id's of some nodes in parent list
+						if enc.getAttr(paCopy) == enc.getAttr(caCopy) {
+							temp := paCopy.next
+							caCopy.next = temp
+							paCopy.next = caCopy
+							exist = true
+							break
+						} else {
+							paCopy = paCopy.next
+						}
+					}
+					if !exist {
+						if unmatchedNodesStartPtr == nil {
+							unmatchedNodesStartPtr = caCopy
+							unmatchedNodesEndPtr = caCopy
+
+						} else {
+							unmatchedNodesEndPtr.next = caCopy
+							unmatchedNodesEndPtr = caCopy
+						}
+					}
+					caCopy = caCopyNext
+					paCopy = tempPa
+				}
+				// Merge all unmatched nodes at the end of parent List
+				temp := tempPa
```

Comment:
Consider this scenario:
// Parent: x d0 n
// Child: d1 d2 d3 d4 d5 n n n n  // Note that child may have duplicate fields as they have come from normalization of nested structures
It can be slightly optimized by directly append [d1 d2 d3 d4 d5] at back of d0 instead of appending one by one.

Reply:
I refactored code. Now we  pick a parent and search in child list and if found put it just before the first occurrence in child list.
So, above scenario won't occur here.